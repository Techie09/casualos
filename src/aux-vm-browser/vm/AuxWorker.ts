export interface WebpackWorker {
    new (): Worker;
}
// // const w = new Worker(new URL('./AuxChannel.worker', import.meta.url)

// load worker by require().
// For some reason, loading relative imports with worker-loader fails when using the import syntax
// but the require syntax works.
const w: WebpackWorker = require('./AuxChannel.worker').default;

export default w;
