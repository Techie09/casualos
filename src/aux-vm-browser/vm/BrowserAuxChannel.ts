import '@casual-simulation/aux-vm/globalThis-polyfill';
import {
    BotAction,
    createAuxPartition,
    PartitionConfig,
    AuxPartition,
} from '@casual-simulation/aux-common';
import { SERVER_ROLE, DeviceAction } from '@casual-simulation/causal-trees';
import { AuxConfig, AuxUser, PortalBundler } from '@casual-simulation/aux-vm';
import { RemoteAuxChannel } from '@casual-simulation/aux-vm-client';
import { createProxyClientPartition } from '../partitions/ProxyClientPartition';
import ESBuildWasmURL from 'esbuild-wasm/esbuild.wasm';

export class BrowserAuxChannel extends RemoteAuxChannel {
    private _defaultHost: string;

    constructor(defaultHost: string, user: AuxUser, config: AuxConfig) {
        super(user, config, {});
        this._defaultHost = defaultHost;
    }

    // TODO: Move this logic to an AuxModule
    // Overridden to automatically execute events from the server.
    protected async _handlePartitionEvents(events: BotAction[]) {
        await super._handlePartitionEvents(events);
        let filtered = events.filter(
            (e) =>
                e.type === 'device' && e.device.roles.indexOf(SERVER_ROLE) >= 0
        ) as DeviceAction[];
        let mapped = <BotAction[]>filtered.map((e) => e.event);
        if (filtered.length > 0) {
            await this.sendEvents(mapped);
        }
    }

    protected async _createPartition(
        config: PartitionConfig
    ): Promise<AuxPartition> {
        let partition = await super._createPartition(config);
        if (!partition) {
            partition = await createAuxPartition(
                config,
                createProxyClientPartition
            );
        }

        return partition;
    }

    protected _createPortalBundler() {
        return new PortalBundler({
            esbuildWasmUrl: new URL(
                ESBuildWasmURL,
                this._config.config.vmOrigin || this._defaultHost
            ).href,
        });
    }
}
