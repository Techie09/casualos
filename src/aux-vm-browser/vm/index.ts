export * from './AuxVMImpl';
export * from './ConnectableAuxVM';
