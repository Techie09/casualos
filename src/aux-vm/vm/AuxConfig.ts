import {
    AuxPartitionConfig,
    AuxDevice,
    BotsState,
    hasValue,
    RemoteCausalRepoProtocol,
} from '@casual-simulation/aux-common';
import { StoredAux } from '../StoredAux';

/**
 * Defines the possible configuration options for a simulation.
 */
export interface AuxConfig {
    config: AuxConfigParameters;

    /**
     * Defines the partitioning structure for bots.
     */
    partitions: AuxPartitionConfig;
}

export interface AuxConfigParameters {
    version: string;
    versionHash: string;
    builder?: string;
    bootstrapState?: StoredAux;
    device?: AuxDevice;

    /**
     * Whether to only allow compiling and runing scripts
     * that have valid signatures.
     */
    forceSignedScripts?: boolean;

    /**
     * Whether to run in debug mode.
     * This will likely cause more verbose console output.
     */
    debug?: boolean;

    /**
     * The connection protocol that causal repo partitions should use.
     */
    causalRepoConnectionProtocol?: RemoteCausalRepoProtocol;

    /**
     * The URL that causal repo partitions should connect to.
     */
    causalRepoConnectionUrl?: string;

    /**
     * The origin that the VM iframe should be loaded from.
     */
    vmOrigin?: string;
}

export function buildVersionNumber(config: AuxConfigParameters) {
    if (!config) {
        return null;
    }
    return {
        hash: config.versionHash,
        ...parseVersionNumber(config.version),
    };
}

/**
 * Parses the given version number.
 * @param version The version number.
 */
export function parseVersionNumber(version: string) {
    if (!version) {
        return {
            version: null,
            major: null,
            minor: null,
            patch: null,
            alpha: null,
        };
    }
    const versionRegex = /^v(\d+)\.(\d+)\.(\d+)(-\w+\.?\d*)?$/i;
    const [str, major, minor, patch, prerelease] = versionRegex.exec(version);

    let alpha: boolean | number = false;
    if (hasValue(prerelease)) {
        alpha = true;
        const [first, number] = prerelease.split('.');
        if (hasValue(number)) {
            alpha = parseInt(number);
        }
    }

    return {
        version: str,
        major: parseInt(major),
        minor: parseInt(minor),
        patch: parseInt(patch),
        alpha,
    };
}
