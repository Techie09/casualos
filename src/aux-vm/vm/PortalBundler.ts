import {
    applyUpdates,
    BotIndex,
    BotIndexEvent,
    BotsState,
    calculateBotValue,
    getScriptPrefix,
    hasPortalScript,
    isPortalScript,
    PrecalculatedBot,
    PrecalculatedBotsState,
    RegisterCustomPortalOptions,
    stateUpdatedEvent,
    StateUpdatedEvent,
    trimPortalScript,
    trimPrefixedScript,
} from '@casual-simulation/aux-common';
import { Observable, Subject } from 'rxjs';
import values from 'lodash/values';
import { isEqual, pick, sortBy } from 'lodash';
import axios from 'axios';
import type ESBuild from 'esbuild';
import * as esbuild from 'esbuild';

export interface PortalEntrypoint {
    botId?: string;
    tag: string;
}

export interface Bundle {
    portalId: string;
    warnings: string[];
    source?: string;
    error?: string;
}

export interface Portal {
    portalId: string;
    entrypoints: PortalEntrypoint[];
    scriptPrefixes: string[];
    style: any;
}

export const DEFAULT_BASE_MODULE_URL: string = 'https://cdn.skypack.dev';

/**
 * Defines a class that is used to bundle scripts for portals.
 * It listens for state updates and is able to asynchrounously emit bundles that should be injected into custom portals.
 */
export class PortalBundler {
    private _esbuildService: ESBuild.Service;
    private _portals: Map<string, Portal>;
    private _state: PrecalculatedBotsState;
    private _onBundleUpdated: Subject<Bundle>;
    private _baseModuleUrl: string = DEFAULT_BASE_MODULE_URL;
    private _httpCache: Map<string, Promise<string>>;
    private _buildCache: Map<string, any>;
    private _index: BotIndex;
    private _esbuildWasmUrl: string;

    /**
     * An observable that emits when a bundle is updated.
     */
    get onBundleUpdated(): Observable<Bundle> {
        return this._onBundleUpdated;
    }

    constructor(options: { esbuildWasmUrl?: string } = {}) {
        this._portals = new Map();
        this._httpCache = new Map();
        this._buildCache = new Map();
        this._index = new BotIndex();
        this._onBundleUpdated = new Subject();
        this._esbuildWasmUrl = options.esbuildWasmUrl || null;
    }

    /**
     * Processes the given state update event.
     */
    stateUpdated(event: StateUpdatedEvent): Promise<void[]> {
        this._state = applyUpdates(this._state, event);
        const batch = this._index.batch(() => {
            const added = event.addedBots.map((id) => this._state[id]);
            const tagUpdates = event.updatedBots
                .map((id) => {
                    let u = event.state[id];
                    let tags = u && u.tags ? Object.keys(u.tags) : [];
                    let valueTags = u && u.values ? Object.keys(u.values) : [];
                    let bot = this._state[id];
                    return {
                        bot,
                        tags: new Set([...tags, ...valueTags]),
                    };
                })
                .filter((u) => !!u.bot);
            if (added.length > 0) {
                this._index.addBots(added);
            }
            if (event.removedBots.length > 0) {
                this._index.removeBots(event.removedBots);
            }
            if (tagUpdates.length > 0) {
                this._index.updateBots(tagUpdates);
            }
        });

        return this._updateBundles(event, batch);
    }

    getPortal(portalId: string) {
        return this._portals.get(portalId);
    }

    /**
     * Registers a custom portal with the given ID.
     * @param portalId The ID of the portal.
     * @param options The options for the portal.
     */
    registerCustomPortal(
        portalId: string,
        options: RegisterCustomPortalOptions
    ): Promise<void> {
        const currentPortal = this._portals.get(portalId);

        if (currentPortal) {
            const prefixesChanged = !isEqual(
                currentPortal.scriptPrefixes,
                options.scriptPrefixes
            );
            const newPortal: Portal = {
                ...currentPortal,
                scriptPrefixes: options.scriptPrefixes,
                style: options.style,
            };

            this._portals.set(portalId, newPortal);

            if (prefixesChanged && newPortal.entrypoints.length > 0) {
                return this._updateBundle(
                    stateUpdatedEvent(this._state),
                    this._index.initialEvents(),
                    newPortal
                );
            }
        } else {
            this._portals.set(portalId, {
                portalId,
                entrypoints: [],
                scriptPrefixes: options.scriptPrefixes,
                style: options.style,
            });
        }

        return Promise.resolve();
    }

    /**
     * Adds an entry point to the portal with the given ID.
     * @param portalId The ID of the portal.
     * @param entrypoint The entrypoint.
     */
    addEntryPoint(
        portalId: string,
        entrypoint: PortalEntrypoint
    ): Promise<void> {
        let portal = this._portals.get(portalId);
        if (portal) {
            portal.entrypoints.push({
                tag: trimPortalScript(portal.scriptPrefixes, entrypoint.tag),
                botId: entrypoint.botId,
            });

            return this._updateBundle(
                stateUpdatedEvent(this._state),
                this._index.initialEvents(),
                portal
            );
        }
    }

    private _updateBundles(
        event: StateUpdatedEvent,
        indexEvents: BotIndexEvent[]
    ) {
        let promises = [] as Promise<void>[];
        for (let portal of this._portals.values()) {
            if (portal.entrypoints.length > 0) {
                promises.push(this._updateBundle(event, indexEvents, portal));
            }
        }
        return Promise.all(promises);
    }

    private _updateBundle(
        event: StateUpdatedEvent,
        indexEvents: BotIndexEvent[],
        portal: Portal
    ) {
        let hasUpdate = indexEvents.some((event) => {
            if (event.type === 'bot_tag_added') {
                return hasPortalScript(
                    portal.scriptPrefixes,
                    calculateBotValue(null, event.bot, event.tag)
                );
            } else if (event.type === 'bot_tag_removed') {
                return hasPortalScript(
                    portal.scriptPrefixes,
                    calculateBotValue(null, event.oldBot, event.tag)
                );
            } else if (event.type === 'bot_tag_updated') {
                const wasEntrypointTag = hasPortalScript(
                    portal.scriptPrefixes,
                    calculateBotValue(null, event.oldBot, event.tag)
                );
                const isEntrypoint = hasPortalScript(
                    portal.scriptPrefixes,
                    calculateBotValue(null, event.bot, event.tag)
                );
                return isEntrypoint || wasEntrypointTag;
            }
        });

        if (hasUpdate) {
            let entryModules = new Set<string>();
            let entryCode = '';
            let bots = sortBy(values(this._state), (b) => b.id);

            for (let entrypoint of portal.entrypoints) {
                let tagModules = bots
                    .map((b) => ({
                        prefix: getScriptPrefix(
                            portal.scriptPrefixes,
                            b.values[entrypoint.tag]
                        ),
                        tag: entrypoint.tag,
                        id: b.id,
                        code: b.values[entrypoint.tag],
                    }))
                    .filter((value) => value.prefix !== null)
                    .map((m) => ({
                        name: auxModuleId(m.prefix, m.id, m.tag),
                        code: trimPrefixedScript(m.prefix, m.code),
                    }));

                for (let m of tagModules) {
                    entryModules.add(m.name);
                }
            }

            for (let name of entryModules) {
                entryCode += `import ${JSON.stringify(name)};\n`;
            }

            return this._esbuild(portal, entryCode, bots);
        }

        return Promise.resolve();
    }

    private async _esbuild(
        portal: Portal,
        entryCode: string,
        bots: PrecalculatedBot[]
    ) {
        if (!this._esbuildService) {
            let options: ESBuild.ServiceOptions = {};
            if (this._esbuildWasmUrl) {
                options.wasmURL = this._esbuildWasmUrl;
            }
            this._esbuildService = await esbuild.startService(options);
        }

        try {
            const result = await this._esbuildService.build({
                entryPoints: ['__entry'],
                bundle: true,
                format: 'iife',
                write: false,
                logLevel: 'silent',
                plugins: [
                    this._esbuildPlugin(portal.scriptPrefixes, entryCode, bots),
                ],
            });

            let final = '';
            for (let file of result.outputFiles) {
                final += file.text;
            }

            const warnings = result.warnings.map((w) => w.text);

            this._onBundleUpdated.next({
                portalId: portal.portalId,
                source: final,
                warnings: warnings,
            });
        } catch (err) {
            this._onBundleUpdated.next({
                portalId: portal.portalId,
                warnings: [],
                error: err.toString(),
            });
        }
    }

    private _esbuildPlugin(
        prefixes: string[],
        entryCode: string,
        bots: PrecalculatedBot[]
    ): ESBuild.Plugin {
        return {
            name: 'casualos',
            setup: (build) => {
                build.onResolve({ filter: /^__entry$/ }, (args) => ({
                    path: args.path,
                    namespace: 'entry-ns',
                }));

                build.onLoad(
                    { filter: /^__entry$/, namespace: 'entry-ns' },
                    (args) => ({
                        contents: entryCode,
                        loader: 'js',
                    })
                );

                build.onResolve({ filter: /\\?auxmodule$/ }, (args) => ({
                    path: args.path,
                    namespace: 'aux-ns',
                }));

                for (let p of prefixes) {
                    let prefix = p;
                    build.onResolve(
                        { filter: new RegExp(`^${prefix}`) },
                        (args) => {
                            const tag = trimPrefixedScript(prefix, args.path);
                            const bot = bots.find((b) =>
                                isPortalScript(prefix, b.values[tag])
                            );

                            if (!bot) {
                                return {
                                    errors: [
                                        {
                                            text: `Unable to resolve "${prefixes[0]}${tag}". No matching script could be found.`,
                                        },
                                    ],
                                };
                            }

                            return {
                                path: auxModuleId(prefix, bot.id, tag),
                                namespace: 'aux-ns',
                            };
                        }
                    );
                }

                build.onLoad(
                    { filter: /\\?auxmodule$/, namespace: 'aux-ns' },
                    (args) => {
                        const { prefix, botId, tag } = parseAuxModuleId(
                            prefixes,
                            args.path
                        );
                        if (prefix && botId && tag) {
                            const bot = this._state[botId];
                            if (!bot) {
                                return {
                                    errors: [
                                        {
                                            text: `Unable to import "${prefix}${tag}". No matching script could be found.`,
                                        },
                                    ],
                                };
                            }
                            const code = bot.values[tag];
                            return {
                                contents: trimPrefixedScript(prefix, code),
                                loader: 'js',
                            };
                        }

                        return {
                            errors: [
                                {
                                    text: `Did you forget to use 📖 when importing?`,
                                },
                            ],
                        };
                    }
                );

                build.onResolve({ filter: /^https?/ }, (args) => ({
                    path: args.path,
                    namespace: 'http-ns',
                }));

                build.onResolve({ filter: /.*/ }, (args) => {
                    const importee = args.path;
                    let importer = args.importer;
                    // convert to HTTP(S) import.
                    if (importee.startsWith('/') || importee.startsWith('./')) {
                        // use importer as base URL
                        if (!importer.endsWith('/')) {
                            importer = importer + '/';
                        }
                        const url = new URL(importee, importer);
                        return { path: url.href, namespace: 'http-ns' };
                    } else {
                        return {
                            path: `${this._baseModuleUrl}/${importee}`,
                            namespace: 'http-ns',
                        };
                    }
                });

                build.onLoad(
                    { filter: /^https?/, namespace: 'http-ns' },
                    async (args) => {
                        try {
                            const cached = this._httpCache.get(args.path);
                            if (typeof cached !== 'undefined') {
                                return {
                                    contents: await cached,
                                    loader: 'js',
                                };
                            }

                            let promise = axios
                                .get(args.path)
                                .then((response) => {
                                    if (typeof response.data === 'string') {
                                        const script = response.data;
                                        return script;
                                    } else {
                                        throw new Error(
                                            `The module server did not return a string.`
                                        );
                                    }
                                });

                            this._httpCache.set(args.path, promise);

                            const script = await promise;
                            return {
                                contents: script,
                                loader: 'js',
                            };
                        } catch (err) {
                            return {
                                errors: [
                                    {
                                        text: `${err}`,
                                    },
                                ],
                            };
                        }
                    }
                );
            },
        };
    }
}

function auxModuleId(prefix: string, botId: string, tag: string) {
    return `${prefix}${botId}.${tag}?auxmodule`;
}

function parseAuxModuleId(
    prefixes: string[],
    id: string
): { prefix: string; botId: string; tag: string } {
    if (isAuxModuleId(id)) {
        for (let prefix of prefixes) {
            if (id.startsWith(prefix)) {
                id = id.substring(prefix.length);
                const dotIndex = id.indexOf('.');
                const botId = id.slice(0, dotIndex);
                const tag = id.slice(
                    dotIndex + 1,
                    id.length - '?auxmodule'.length
                );

                return {
                    prefix,
                    botId,
                    tag,
                };
            }
        }
    }
    return { prefix: null, botId: null, tag: null };
}

function isAuxModuleId(id: string) {
    return id.endsWith('?auxmodule');
}
