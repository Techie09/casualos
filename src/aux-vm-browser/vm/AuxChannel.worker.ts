import '@casual-simulation/aux-vm/globalThis-polyfill';
import { expose } from 'comlink';
import { listenForChannel } from '../html/IFrameHelpers';
import { BrowserAuxChannel } from './BrowserAuxChannel';
import { NestedWorker } from './NestedWorker';

// Use the nested worker polyfill if the worker interface is not specified.
if (!globalThis.Worker) {
    const nativePostMessage = postMessage.bind(self);
    (<any>globalThis).Worker = class extends NestedWorker {
        constructor(stringOrUrl: string | URL, options?: WorkerOptions) {
            super(nativePostMessage, stringOrUrl, options);
        }
    };
}

listenForChannel().then((port) => {
    console.log('[AuxChannel.worker] Got port, exposing API');
    expose(BrowserAuxChannel, port);
});

console.log('[AuxChannel.worker] Listening for port...');
