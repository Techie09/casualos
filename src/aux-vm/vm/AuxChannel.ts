import {
    LocalActions,
    BotAction,
    StateUpdatedEvent,
    RuntimeStateVersion,
    RegisterCustomPortalAction,
} from '@casual-simulation/aux-common';
import {
    StatusUpdate,
    DeviceAction,
    CurrentVersion,
    VersionVector,
} from '@casual-simulation/causal-trees';
import { AuxConfig } from './AuxConfig';
import { AuxChannelErrorType } from './AuxChannelErrorTypes';
import { AuxUser } from '../AuxUser';
import { Observable, SubscriptionLike } from 'rxjs';
import { StoredAux } from '../StoredAux';
import { AuxVM } from './AuxVM';
import { PortalEvent } from './PortalEvents';

/**
 * Defines an interface for the static members of an AUX.
 */
export interface AuxStatic {
    /**
     * Creates a new AUX using the given config.
     */
    new (defaultHost: string, user: AuxUser, config: AuxConfig): AuxChannel;
}

/**
 * Defines an interface for an AUX.
 * That is, a channel that interfaces with the AUX bot format in realtime.
 */
export interface AuxChannel extends SubscriptionLike {
    /**
     * The observable that should be triggered whenever a device event is sent to the AUX.
     */
    onDeviceEvents: Observable<DeviceAction[]>;

    /**
     * The observable that should be triggered whenever a local event is emitted from the AUX.
     */
    onLocalEvents: Observable<LocalActions[]>;

    /**
     * The observable that should be triggered whenever the bots state is updated.
     */
    onStateUpdated: Observable<StateUpdatedEvent>;

    /**
     * The observable that should be triggered whenever the state version updated.
     */
    onVersionUpdated: Observable<RuntimeStateVersion>;

    /**
     * The observable that should be triggered whenever the connection state changes.
     */
    onConnectionStateChanged: Observable<StatusUpdate>;

    /**
     * The observable that should be triggered whenever a portal event occurs.
     */
    onPortalEvents: Observable<PortalEvent[]>;

    /**
     * The observable that is resolved whenever an error occurs.
     */
    onError: Observable<AuxChannelErrorType>;

    /**
     * Initializes the AUX.
     * @param onLocalEvents The callback that should be triggered whenever a local event is emitted from the AUX.
     * @param onDeviceEvents The callback that should be triggered whenever a device event it emitted from the AUX.
     * @param onStateUpdated The callback that should be triggered whenever the bots state is updated.
     * @param onConnectionStateChanged The callback that should be triggered whenever the connection state changes.
     * @param onPortalEvents The callback that should be triggered whenever a portal event occurs.
     * @param onError The callback that should be triggered whenever an error occurs.
     */
    init(
        onLocalEvents?: (events: LocalActions[]) => void,
        onDeviceEvents?: (events: DeviceAction[]) => void,
        onStateUpdated?: (state: StateUpdatedEvent) => void,
        onVersionUpdated?: (version: RuntimeStateVersion) => void,
        onConnectionStateChanged?: (state: StatusUpdate) => void,
        onPortalEvents?: (events: PortalEvent[]) => void,
        onError?: (err: AuxChannelErrorType) => void
    ): Promise<void>;

    /**
     * Initializes the AUX and waits for the connection to be initialized.
     * @param onLocalEvents The callback that should be triggered whenever a local event is emitted from the AUX.
     * @param onDeviceEvents The callback that should be triggered whenever a device event it emitted from the AUX.
     * @param onStateUpdated The callback that should be triggered whenever the bots state is updated.
     * @param onConnectionStateChanged The callback that should be triggered whenever the connection state changes.
     * @param onPortalEvents The callback that should be triggered whenever a portal event occurs.
     * @param onError The callback that should be triggered whenever an error occurs.
     */
    initAndWait(
        onLocalEvents?: (events: LocalActions[]) => void,
        onDeviceEvents?: (events: DeviceAction[]) => void,
        onStateUpdated?: (state: StateUpdatedEvent) => void,
        onVersionUpdated?: (version: RuntimeStateVersion) => void,
        onConnectionStateChanged?: (state: StatusUpdate) => void,
        onPortalEvents?: (events: PortalEvent[]) => void,
        onError?: (err: AuxChannelErrorType) => void
    ): Promise<void>;

    /**
     * Sets the user that the channel should use.
     * @param user The user.
     */
    setUser(user: AuxUser): Promise<void>;

    /**
     * Sets the grant that the channel should use to authenticate the user.
     * @param grant The grant to use.
     */
    setGrant(grant: string): Promise<void>;

    /**
     * Sends the given list of bots events to the AUX for processing.
     * @param events The events.
     */
    sendEvents(events: BotAction[]): Promise<void>;

    /**
     * Executes a shout with the given event name on the given bot IDs with the given argument.
     * Also dispatches any actions and errors that occur.
     * Returns the results from the event.
     * @param eventName The name of the event.
     * @param botIds The IDs of the bots that the shout is being sent to.
     * @param arg The argument to include in the shout.
     */
    shout(
        eventName: string,
        botIds?: string[],
        arg?: any
    ): Promise<ChannelActionResult>;

    /**
     * Runs the given list of formulas.
     * @param formulas The formulas.
     */
    formulaBatch(formulas: string[]): Promise<void>;

    /**
     * Forks the AUX into the channel with the given ID.
     * @param newId The ID that the new AUX should have.
     */
    forkAux(newId: string): Promise<void>;

    /**
     * Exports the atoms for the given bots.
     * @param botIds The bots to export.
     */
    exportBots(botIds: string[]): Promise<StoredAux>;

    /**
     * Exports the causal tree for the simulation.
     */
    export(): Promise<StoredAux>;

    /**
     * Gets the list of tags that are in use.
     */
    getTags(): Promise<string[]>;
}

export interface ChannelActionResult {
    /**
     * The actions that were queued.
     */
    actions: BotAction[];

    /**
     * The results from the scripts that were run.
     */
    results: any[];
}
