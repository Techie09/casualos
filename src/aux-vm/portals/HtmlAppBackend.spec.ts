import {
    asyncResult,
    AuxPartitions,
    AuxRuntime,
    BotAction,
    botAdded,
    createBot,
    createMemoryPartition,
    htmlAppEvent,
    iteratePartitions,
    MemoryPartition,
    ON_APP_SETUP_ACTION_NAME,
    registerHtmlApp,
    setAppOutput,
    stateUpdatedEvent,
    toast,
    updateHtmlApp,
    UpdateHtmlAppAction,
} from '@casual-simulation/aux-common';
import { waitAsync } from '@casual-simulation/aux-common/test/TestHelpers';
import { Subscription } from 'rxjs';
import { HtmlAppBackend } from './HtmlAppBackend';
import { v4 as uuid } from 'uuid';
import { AuxHelper } from '../vm';
import { skip, tap } from 'rxjs/operators';
import { h } from 'preact';
import htm from 'htm';

const uuidMock: jest.Mock = <any>uuid;
jest.mock('uuid');

describe('HtmlAppBackend', () => {
    let runtime: AuxRuntime;
    let actions: BotAction[];
    let memory: MemoryPartition;
    let userId: string = 'user';
    let sub: Subscription;

    beforeEach(async () => {
        actions = [];
        sub = new Subscription();
        runtime = new AuxRuntime(
            {
                hash: 'hash',
                major: 1,
                minor: 0,
                patch: 0,
                version: 'v1.0.0',
            },
            {
                supportsAR: false,
                supportsVR: false,
                isCollaborative: true,
                ab1BootstrapUrl: 'ab1Bootstrap',
            }
        );
        memory = createMemoryPartition({
            type: 'memory',
            initialState: {},
        });
        memory.space = 'shared';

        await memory.applyEvents([botAdded(createBot('user'))]);
    });

    function createHelper(partitions: AuxPartitions) {
        runtime = new AuxRuntime(
            {
                hash: 'hash',
                major: 1,
                minor: 0,
                patch: 0,
                version: 'v1.0.0',
            },
            {
                supportsAR: false,
                supportsVR: false,
                isCollaborative: true,
                ab1BootstrapUrl: 'ab1Bootstrap',
            }
        );
        const helper = new AuxHelper(partitions, runtime);

        for (let [, partition] of iteratePartitions(partitions)) {
            sub.add(
                partition.onStateUpdated
                    .pipe(
                        tap((e) => {
                            runtime.stateUpdated(e);
                        })
                    )
                    .subscribe(null, (e: any) => console.error(e))
            );
        }

        runtime.userId = userId;
        sub.add(helper.localEvents.subscribe((a) => actions.push(...a)));

        return helper;
    }

    afterEach(() => {
        sub.unsubscribe();
    });

    it('should send a register_html_app event when created', async () => {
        const helper = createHelper({
            shared: memory,
        });
        uuidMock.mockReturnValueOnce('uuid');

        let portal = new HtmlAppBackend('testPortal', 'myBot', helper);

        await waitAsync();

        expect(actions).toEqual([registerHtmlApp('testPortal', 'uuid')]);
    });

    describe('onAppSetup', () => {
        it('should send a onRender action when the register_html_app result is returned', async () => {
            const helper = createHelper({
                shared: memory,
            });
            await helper.transaction(
                botAdded(
                    createBot('myBot', {
                        [ON_APP_SETUP_ACTION_NAME]: `@tags.rendered = true`,
                        rendered: false,
                    })
                )
            );

            uuidMock.mockReturnValueOnce('uuid');

            let portal = new HtmlAppBackend('testPortal', 'myBot', helper);

            await waitAsync();

            expect(helper.botsState['myBot'].tags.rendered).toBe(false);

            portal.handleEvents([asyncResult('uuid', null)]);

            await waitAsync();

            expect(runtime.currentState['myBot'].values.rendered).toBe(true);
        });

        it('should include the HTML document in the onAppSetup action', async () => {
            const helper = createHelper({
                shared: memory,
            });
            await helper.transaction(
                botAdded(
                    createBot('myBot', {
                        [ON_APP_SETUP_ACTION_NAME]: `@that.document.body.appendChild(that.document.createElement('h1'))`,
                    })
                )
            );

            uuidMock.mockReturnValueOnce('uuid');

            let portal = new HtmlAppBackend('testPortal', 'myBot', helper);

            await waitAsync();

            portal.handleEvents([asyncResult('uuid', null)]);

            await waitAsync();

            expect(actions.slice(1)).toEqual([
                registerHtmlApp('testPortal', 'uuid'),
                updateHtmlApp('testPortal', [
                    {
                        type: 'childList',
                        target: expect.objectContaining({
                            __id: '0',
                        }),
                        addedNodes: [
                            expect.objectContaining({
                                __id: '1',
                                nodeName: 'H1',
                            }),
                        ],
                        removedNodes: [],
                        attributeName: undefined,
                        attributeNamespace: undefined,
                        nextSibling: undefined,
                        previousSibling: undefined,
                        oldValue: undefined,
                    },
                ]),
            ]);
        });

        it('should call event listeners for the html_app_event events', async () => {
            const helper = createHelper({
                shared: memory,
            });
            await helper.transaction(
                botAdded(
                    createBot('myBot', {
                        [ON_APP_SETUP_ACTION_NAME]: `@let h1 = that.document.createElement('h1'); h1.addEventListener('click', () => tags.clicked = true); that.document.body.appendChild(h1);`,
                        clicked: false,
                    })
                )
            );

            uuidMock.mockReturnValueOnce('uuid');

            let portal = new HtmlAppBackend('testPortal', 'myBot', helper);

            await waitAsync();

            portal.handleEvents([asyncResult('uuid', null)]);

            await waitAsync();

            portal.handleEvents([
                htmlAppEvent('testPortal', {
                    type: 'click',
                    target: '1',
                }),
            ]);

            await waitAsync();

            expect(runtime.currentState['myBot'].values.clicked).toBe(true);
        });

        it('should emit an async result for the given register task ID', async () => {
            const helper = createHelper({
                shared: memory,
            });
            await helper.transaction(
                botAdded(
                    createBot('myBot', {
                        [ON_APP_SETUP_ACTION_NAME]: `@os.toast("Hit")`,
                    })
                )
            );

            uuidMock.mockReturnValueOnce('uuid');

            let portal = new HtmlAppBackend(
                'testPortal',
                'myBot',
                helper,
                'taskId'
            );

            await waitAsync();

            portal.handleEvents([asyncResult('uuid', null)]);

            await waitAsync();

            expect(actions.slice(2)).toEqual([
                toast('Hit'),
                asyncResult('taskId', null),
            ]);
        });
    });

    describe('set_app_output', () => {
        it('should render the output as a preact component', async () => {
            const helper = createHelper({
                shared: memory,
            });

            uuidMock.mockReturnValueOnce('uuid');

            let portal = new HtmlAppBackend('testPortal', 'myBot', helper);

            await waitAsync();

            const html = htm.bind(h);

            portal.handleEvents([asyncResult('uuid', null)]);
            portal.handleEvents([
                setAppOutput('testPortal', html`<h1>Hello</h1>`),
            ]);

            await waitAsync();

            expect(actions.length).toBe(2);
            expect(actions[0]).toEqual(registerHtmlApp('testPortal', 'uuid'));

            const updateAction = actions[1] as UpdateHtmlAppAction;

            expect(updateAction).toMatchSnapshot();

            expect(updateAction.type).toBe('update_html_app');
            expect(updateAction.appId).toBe('testPortal');
            expect(updateAction.updates.length).toBe(1);
            expect(updateAction.updates[0].type).toBe('childList');
            expect(updateAction.updates[0].addedNodes.length).toBe(1);
            const h1Node = updateAction.updates[0].addedNodes[0] as any;
            expect(h1Node.nodeName).toBe('H1');
            expect(h1Node.childNodes.length).toBe(1);
            expect(h1Node.childNodes[0].nodeName).toBe('#text');
        });
    });
});
