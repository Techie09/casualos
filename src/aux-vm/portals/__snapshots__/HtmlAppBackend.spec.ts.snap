// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HtmlAppBackend set_app_output should render the output as a preact component 1`] = `
Object {
  "appId": "testPortal",
  "type": "update_html_app",
  "updates": Array [
    Object {
      "addedNodes": Array [
        Object {
          "__id": "1",
          "attributes": Array [],
          "childNodes": Array [
            Object {
              "__id": "2",
              "childNodes": Array [],
              "data": "Hello",
              "nodeName": "#text",
              "nodeType": 3,
            },
          ],
          "nodeName": "H1",
          "nodeType": 1,
          "style": Object {},
        },
      ],
      "nextSibling": undefined,
      "previousSibling": undefined,
      "removedNodes": Array [],
      "target": Object {
        "__id": "0",
        "attributes": Array [],
        "childNodes": Array [
          Object {
            "__id": "1",
            "attributes": Array [],
            "childNodes": Array [
              Object {
                "__id": "2",
                "childNodes": Array [],
                "data": "Hello",
                "nodeName": "#text",
                "nodeType": 3,
              },
            ],
            "nodeName": "H1",
            "nodeType": 1,
            "style": Object {},
          },
        ],
        "nodeName": "BODY",
        "nodeType": 1,
        "style": Object {},
      },
      "type": "childList",
    },
  ],
}
`;
